django/apps/config.py

```diff
@@ -82,87 +87,157 @@ def _path_from_module(self, module):
+                        raise RuntimeError(
+                            "%r declares more than one default AppConfig: "
+                            "%s." % (mod_path, ", ".join(candidates))
+                        )
+                    elif len(app_configs) == 1:
+                        app_config_class = app_configs[0][1]
+                        app_config_name = '%s.%s' % (mod_path, app_configs[0][0])
+
+            # If app_module specifies a default_app_config, follow the link.
+            # default_app_config is deprecated, but still takes over the
+            # automatic detection for backwards compatibility during the
+            # deprecation period.
             try:
-                # If this works, the app module specifies an app config class.
-                entry = module.default_app_config
+                new_entry = app_module.default_app_config
             except AttributeError:
-                # Otherwise, it simply uses the default app config class.
-                return cls(entry, module)
+                # Use the default app config class if we didn't find anything.
+                if app_config_class is None:
+                    app_config_class = cls
+                    app_name = entry
             else:
-                mod_path, _, cls_name = entry.rpartition('.')
-
-        # If we're reaching this point, we must attempt to load the app config
-        # class located at <mod_path>.<cls_name>
-        mod = import_module(mod_path)
-        try:
-            cls = getattr(mod, cls_name)
-        except AttributeError:
-            if module is None:
-                # If importing as an app module failed, check if the module
-                # contains any valid AppConfigs and show them as choices.
-                # Otherwise, that error probably contains the most informative
-                # traceback, so trigger it again.
-                candidates = sorted(
-                    repr(name) for name, candidate in mod.__dict__.items()
-                    if isinstance(candidate, type) and
-                    issubclass(candidate, AppConfig) and
-                    candidate is not AppConfig
+                message = (
+                    "%r defines default_app_config = %r. "
+                    % (entry, new_entry)
                 )
-                if candidates:
-                    raise ImproperlyConfigured(
-                        "'%s' does not contain a class '%s'. Choices are: %s."
-                        % (mod_path, cls_name, ', '.join(candidates))
+                if new_entry == app_config_name:
+                    message += (
+                        'Django now detects this configuration automatically. '
+                        'You can remove default_app_config.'
                     )
-                import_module(entry)
+                else:
+                    message += (
+                        "Django detects another configuration automatically, "
```

Comment:
`Django detects another configuration automatically` reads awkwardly to me. I assume it should be taken to mean Django detected another candidate to use as the default?

Or put another way - the test case (elsewhere in the PR) which shows the full output of this warning being raised doesn't tell me (as someone seeing it for the first time) what the _actual_ problem is, nor what I should to do to fix it (beyond review my configuration)

Reply:
This is a highly improbable edge case:

- app contains an `apps` submodule declaring a single `AppConfig` subclass
- app `__init__.py` contains a `default_app_config` variable pointing to another `AppConfig` subclass

I added this check and error message for the sake of completeness but I'm failing to imagine a legitimate use case for this.

Thanks for pointing out that the message isn't sufficiently helpful. I agree that we should do better. How about this?

```
"However, Django's automatic detection picked another configuration, "
"%r. You should move the default AppConfig class to the apps submodule "
" of your application and, if this module defines several AppConfig classes, "
"mark the default one with default = True."
```

Reply:
I like that it's a highly improbable edge case but you've handled it anyway :)
Yep that text (or something like it) seems clearer to me, as it's giving instruction for resolution, should it ever happen to anyone.